src/rgw/rgw_bucket_snap_types.h

```diff
@@ -0,0 +1,171 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#pragma once
+
+#include <string>
+#include <map>
+#include "include/types.h"
+#include "common/Formatter.h"
+#include "common/ceph_time.h"
+#include "common/ceph_json.h"
+#include "rgw_xml.h"
+
+
+
+struct rgw_bucket_snap_id {
+  static constexpr uint64_t SNAP_UNDEFINED = (uint64_t)0;
+  static constexpr uint64_t SNAP_MIN       = (uint64_t)1;
+  uint64_t snap_id;
+
+  constexpr rgw_bucket_snap_id() : snap_id(SNAP_UNDEFINED) {}
+  rgw_bucket_snap_id(uint64_t snap_id) : snap_id(snap_id) {}
+
+  std::string to_string() const {
+    return std::to_string(snap_id);
+  }
+
+  void init() {
```

Comment:
I don't know all the code yet, but could initialization method just be a constructor?

Reply:
@mattbenjamin no, that's on purpose. Implicitly it'll be set to UNDEFINED by the constructor.

Reply:
yeah the constructors look good. usually would expect `operator=` for these re-init cases, where a single assignment operator could replace all 3 of these `init()` overloads:
```c++
snapid = rgw_bucket_snap_id(); // init()
snapid = other_snapid; // init(rgw_bucket_snap_id)
snapid = 12; // init(uint64_t) and relies on non-explicit ctor rgw_bucket_snap_id(uint64_t)
```

Reply:
right, I kept the init() though, we can revise later